flecsi/exec/hpx/task_prologue.hh

```diff
@@ -0,0 +1,222 @@
+   /@@@@@@@  /@@ @@///@@/@@       /@@@@@@@@@/@@
+   /@@////   /@@/@@@@@@@/@@       ////////@@/@@
+   /@@       /@@/@@//// //@@    @@       /@@/@@
+   /@@       @@@//@@@@@@ //@@@@@@  @@@@@@@@ /@@
+   //       ///  //////   //////  ////////  //
+
+   Copyright (c) 2016, Triad National Security, LLC
+   All rights reserved.
+                                                                              */
+
+#ifndef FLECSI_EXEC_HPX_TASK_PROLOGUE_HH
+#define FLECSI_EXEC_HPX_TASK_PROLOGUE_HH
+
+/*! @file */
+
+#include <flecsi-config.h>
+
+#if !defined(FLECSI_ENABLE_HPX)
+#error FLECSI_ENABLE_HPX not defined! This file depends on HPX!
+#endif
+
+#include <hpx/modules/collectives.hpp>
+#include <hpx/modules/concurrency.hpp>
+#include <hpx/modules/serialization.hpp>
+
+#include "flecsi/data/privilege.hh"
+#include "flecsi/data/topology.hh"
+#include "flecsi/exec/hpx/future.hh"
+#include "flecsi/flog.hh"
+#include "flecsi/util/demangle.hh"
+
+#include <memory>
+#include <mutex>
+#include <string>
+
+namespace flecsi {
+namespace topo {
+struct global_base;
+}
+
+namespace exec {
+
+template<task_processor_type_t ProcessorType>
+struct task_prologue {
+
+protected:
+  // Those methods are "protected" because they are *only* called by
+  // flecsi::exec::prolog() which inherits from task_prologue.
+
+  // Patch up the "un-initialized" type conversion from future<R, index>
+  // to future<R, single> in the generic code.
+  template<typename R>
+  static void visit(future<R, exec::launch_type_t::single> & single,
+    future<R, exec::launch_type_t::index> & index) {
+    single = index.get(flecsi::run::context::instance().color());
+  }
+
+  template<typename T,
+    Privileges P,
+    class Topo,
+    typename Topo::index_space Space>
+  void visit(data::accessor<data::raw, T, P> &,
+    const data::field_reference<T, data::raw, Topo, Space> & ref) {
+
+    // The spinlock-pool is used to protect the initialization of the
+    // futures that are associated with the storage for a field. While
+    // all operations on the future itself are thread-safe, assigning
+    // a future is not thread-safe.
+    using spinlock_pool =
+      ::hpx::util::spinlock_pool<::hpx::shared_future<void>>;
+
+    const field_id_t f = ref.fid();
+    constexpr bool glob =
+      std::is_same_v<typename Topo::base, topo::global_base>;
+
+    auto & r = [&]() -> auto & {
+      auto & t = ref.topology();
+      if constexpr(glob) {
+        return *t;
+      }
+      else {
+        // The partition controls how much memory is allocated.
+        return t.template get_partition<Space>(f);
+      }
+    }
+    ();
+
+    // If the current argument has a valid future associated with it, then
+    // it has to be used a dependency for executing the current task.
+    auto & field_future = r[f].future;
+    if(field_future.valid()) {
+      std::lock_guard l(spinlock_pool::spinlock_for(&field_future));
+      dependencies.push_back(field_future);
```

Comment:
Recalling that `task_prologue` is never used concurrently (in one process), do we need to protect this access?  The task uses the promise to make the future ready without actually interacting with `backend_storage::future`, doesn't it?

Reply:
Yes, that's what I was thinking today as well. I think we can remove those lock_guards.